cpp/src/svm/smoblocksolve.h

```diff
@@ -0,0 +1,246 @@
+ * \f$\Delta\alpha\f$.
+ *
+ * In every iteration we select the two vectors using the following formulas
+ *   \f[ u = \mathrm{argmin}_{i=1}^{n_{ws}}\left[ f_i |
+ *      x_i \in X_\mathrm{upper} \right] \f]
+ *
+ *  \f[ l = \mathrm{argmax}_{i=1}^{n_{ws}} \left[
+ *          \frac{(f_u-f_i)^2}{\eta_i}| f_u < f_i \land x_i \in
+ *           X_{\mathrm{lower}}\right], \f]
+ *  where \f[\eta_i = K(x_u, x_u) + K(x_i, x_i) - 2K(x_u, x_i). \f]
+ *
+ * We update the values of the dual coefs according to (additionally we clip
+ * values so that the coefficients stay in the [0, C] interval)
+ * \f[ \Delta \alpha_l = - y_l \frac{f_l - f_u}{\eta_l} = -y_l q, \f]
+ * \f[ \alpha_l += \Delta \alpha_l, \f]
+ * \f[ \Delta \alpha_u = -y_u y_l \Delta \alpha_l = y_u q, \f]
+ * \f[ \alpha_u += \Delta \alpha_u. \f]
+ *
+ * We also update the optimality indicator vector for the WS:
+ * \f[ f_i += \Delta\alpha_u y_u K(x_u,x_i) + \Delta\alpha_l y_l K(x_l, x_i) \f]
+ *
+ * During the inner iterations, the f values are updated only for the WS
+ * (since we are solving the optimization subproblem for the WS subproblem).
+ * For consistency, f is kept as an input parameter, the changed values are
+ * not saved back to global memory. After this solver finishes,  all the f
+ * values (WS and outside WS) f should be updated using the delta_alpha output
+ * parameter.
+ *
+ * References:
+ * - [1] J. C. Platt Sequential Minimal Optimization: A Fast Algorithm for
+ *      Training Support Vector Machines, Technical Report MS-TR-98-14 (1998)
+ * - [2] S.S. Keerthi et al. Improvements to Platt's SMO Algorithm for SVM
+ *      Classifier Design, Neural Computation 13, 637-649 (2001)
+ * - [3] Z. Wen et al. ThunderSVM: A Fast SVM Library on GPUs and CPUs, Journal
+ *      of Machine Learning Research, 19, 1-5 (2018)
+ *
+ * @tparam math_t floating point data type
+ * @tparam WSIZE working set size (max 1024)
+ * @param [in] y_array target labels size [n_rows]
+ * @param [in] n_rows number of trainig vectors
+ * @param [inout] alpha dual coefficients, size [n_rows]
+ * @param [in] n_ws number of elements in the working set
+ * @param [out] delta_alpha change in the dual coeff of vectors in the working
+ *        set, size [n_ws]
+ * @param [in] f_array optimality indicator vector, size [n_rows]
+ * @param [in] kernel kernel function calculated between the working set and all
+ *   other training vectors, size [n_rows * n_ws]
+ * @param [in] ws_idx indices of traning vectors in the working set, size [n_ws]
+ * @param [in] C penalty parameter
+ * @param [in] eps tolerance, iterations will stop if the duality gap is smaller
+ *  than this value (or if the gap is smaller than 0.1 times the initial gap)
+ * @param [out] return_buff, two valies are returned: duality gap and the number
+ *   of iterations
+ * @param [in] max_iter maximum number of iterations
+ */
+template <typename math_t, int WSIZE>
+__global__ void SmoBlockSolve(math_t *y_array, int n_rows, math_t *alpha,
```

Comment:
If there is any way, I'd split this long kernel into multiple pieces. Not necessarily, but usually, these long kernels end up getting serialized due to warp divergence. Also, it's not easy to do coalesced access to memory.

Reply:
There is some divergence, but we should be far from being serialized. Coalesced memory access is hampered by indirect indexing of the kernel array using idx, other accesses (Kd) are through shared memory.

For performance reasons the inner loop should be a single kernel. For readability I agree that it would be better to cut it into smaller pieces which could be inlined. I would do it post 0.10, created issue #1141.